Improve signing of certificate

HanslettTheDev opened the pull request:
## Summary

This pull requests aims to improve the work done in [#113](https://github.com/PyAr/pyafipws/pull/113). Updating the cryptography module, fixing the broken tests and libraries

## Checklist

- [x] Classes, Variables, function and methods logic  ok
- [x] Comments written explaining what the code does
- [x] Test coverage with pytest implemented

HanslettTheDev committed:
bump: this commit is a copy of #113 by saliticus to change the signing method

Signed-off-by: HanslettTheDev <HanslettTheDev@gmail.com>

HanslettTheDev committed:
feat: backward compatible signing for python2.7

Signed-off-by: HanslettTheDev <HanslettTheDev@gmail.com>

HanslettTheDev committed:
bump: cryptography update 39.0.2 -> 41.0.1

Signed-off-by: HanslettTheDev <HanslettTheDev@gmail.com>

restart actions

Signed-off-by: HanslettTheDev <HanslettTheDev@gmail.com>

bug: added a bug to find an existing issue

Signed-off-by: HanslettTheDev <HanslettTheDev@gmail.com>

undo-bug: removed the bug

Signed-off-by: HanslettTheDev <HanslettTheDev@gmail.com>

fix: updated the wsaa.py file

Signed-off-by: HanslettTheDev <HanslettTheDev@gmail.com>

bump: cryptography update 39.0.2 -> 41.0.1

HanslettTheDev committed:
bump: cryptography update 39.0.2 -> 41.0.1

Signed-off-by: HanslettTheDev <HanslettTheDev@gmail.com>

HanslettTheDev committed:
bump: cryptography update 39.0.2 -> 41.0.1

Signed-off-by: HanslettTheDev <HanslettTheDev@gmail.com>

HanslettTheDev committed:
restart actions

Signed-off-by: HanslettTheDev <HanslettTheDev@gmail.com>

HanslettTheDev committed:
bug: added a bug to find an existing issue

Signed-off-by: HanslettTheDev <HanslettTheDev@gmail.com>

HanslettTheDev committed:
undo-bug: removed the bug

Signed-off-by: HanslettTheDev <HanslettTheDev@gmail.com>

HanslettTheDev committed:
fix: updated the wsaa.py file

Signed-off-by: HanslettTheDev <HanslettTheDev@gmail.com>

HanslettTheDev committed:
Merge branch 'fix_signature' of https://github.com/HanslettTheDev/pyafipws into fix_signature

NicolasSandoval commented on wsaa.py:
```diff
@@ -116,6 +117,7 @@ def sign_tra(tra, cert=CERT, privatekey=PRIVATEKEY, passphrase=""):
         # Crear un buffer desde el texto
+        # Se crea un buffer nuevo porque la firma lo consume
         bio_in = _lib.BIO_new_mem_buf(tra, len(tra))
 
```
I think creating a memory buffer for the signing process, including all the structure like `_lib` and `_ffi` could go with the rest of the code handled by python2. The handling of the certificates is fine, which is out of the conditional.

HanslettTheDev replied:
Done! Thanks

NicolasSandoval commented on wsaa.py:
```diff
@@ -136,42 +138,62 @@ def sign_tra(tra, cert=CERT, privatekey=PRIVATEKEY, passphrase=""):
-        try:
-            # Crear buffer de salida
-            bio_out = _lib.BIO_new(_lib.BIO_s_mem())
+        if sys.version.split(" ")[0][0:3] == "2.7":
```
Maybe you would only check the major version number.
Something like:
```suggestion
        MAJOR = 0 # first string element
        if sys.version[MAJOR] == "2":
```
or use the `version_info` method
```suggestion
        if sys.version_info.major == 2:
```

NicolasSandoval commented on wsaa.py:
```diff
@@ -136,42 +138,62 @@ def sign_tra(tra, cert=CERT, privatekey=PRIVATEKEY, passphrase=""):
+                _lib.BIO_free(bio_in)
+
+            # Generar p7 en formato mail y recortar headers
+            msg = email.message_from_string(output.decode("utf8"))
+            for part in msg.walk():
+                filename = part.get_filename()
+                if filename == "smime.p7m":
+                    # Es la parte firmada?
+                    # Devolver CMS
+                    return part.get_payload(decode=False)
```
Would you try here to remove the repeated code? You could use `else` to separate the python3 code. I think it's better to rename `output` to `p7` directly.

HanslettTheDev replied:
Done! Changes made! This was a cool idea to reduce the repeated code. I was worried about this too

NicolasSandoval commented on wsaa.py:
```diff
@@ -136,42 +138,62 @@ def sign_tra(tra, cert=CERT, privatekey=PRIVATEKEY, passphrase=""):
         for part in msg.walk():
             filename = part.get_filename()
-            if filename == "smime.p7m":
+            if filename == "smime.p7s":
```
One possibility to avoid the `m` and `s` problem is to use the `startswith` method.
For instance:
```suggestion
            if filename and filename.startswith("smime.p7"):
```

NicolasSandoval reviewed (changes requested):
I left some possibilities for small structural changes. Good job @HanslettTheDev.

HanslettTheDev committed:
refactor: used the sys module to check for the major version

Co-authored-by: Nico Sandoval <49107024+NicolasSandoval@users.noreply.github.com>

HanslettTheDev committed:
Merge branch 'main' into fix_signature

HanslettTheDev committed:
refactor: removed repeated code in signing certificates

Signed-off-by: HanslettTheDev <HanslettTheDev@gmail.com>

HanslettTheDev committed:
bump: cryptography update 39.0.2 -> 41.0.1

Signed-off-by: HanslettTheDev <HanslettTheDev@gmail.com>

restart actions

Signed-off-by: HanslettTheDev <HanslettTheDev@gmail.com>

bug: added a bug to find an existing issue

Signed-off-by: HanslettTheDev <HanslettTheDev@gmail.com>

undo-bug: removed the bug

Signed-off-by: HanslettTheDev <HanslettTheDev@gmail.com>

fix: updated the wsaa.py file

Signed-off-by: HanslettTheDev <HanslettTheDev@gmail.com>

refactor: used the sys module to check for the major version

Co-authored-by: Nico Sandoval <49107024+NicolasSandoval@users.noreply.github.com>

Only run release on original repository, not forks

This avoids errors with dependabot PR:
```
Run marvinpinto/action-automatic-releases@latest
Initializing the Automatic Releases action
Determining release tags
Retrieving commit history
Generating changelog
Generating release tag
  Attempting to create or update release tag "beta"
  Could not create new tag "refs/tags/beta" (Resource not accessible by integration) therefore updating existing tag "tags/beta"
  Error: Resource not accessible by integration
```

Signed-off-by: Nicolás Sandoval <nicolassandovalunm@gmail.com>

Properly fix dependabot read-only token

https://github.com/dependabot/dependabot-core/issues/3253
Signed-off-by: Nicolás Sandoval <nicolassandovalunm@gmail.com>

Bump future from 0.18.2 to 0.18.3

Bumps [future](https://github.com/PythonCharmers/python-future) from 0.18.2 to 0.18.3.
- [Release notes](https://github.com/PythonCharmers/python-future/releases)
- [Changelog](https://github.com/PythonCharmers/python-future/blob/master/docs/changelog.rst)
- [Commits](https://github.com/PythonCharmers/python-future/compare/v0.18.2...v0.18.3)

---
updated-dependencies:
- dependency-name: future
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Signed-off-by: Nicolás Sandoval <nicolassandovalunm@gmail.com>

Cumulative updates f/ develop (python3 conversion)

* 2to3 -w file.py
* python -m black file.py
* iconv -f iso-8859-1 -t utf-8 < file_new.py > file.py

Signed-off-by: Nicolás Sandoval <nicolassandovalunm@gmail.com>

Update copyright and remove merge typo

Signed-off-by: Nicolás Sandoval <nicolassandovalunm@gmail.com>

WSFEv1: update copyright

Signed-off-by: Nicolás Sandoval <nicolassandovalunm@gmail.com>

WSFEv1: re-sync updates

Signed-off-by: Nicolás Sandoval <nicolassandovalunm@gmail.com>

WSFEXv1: disable get parameters until UT updated

Signed-off-by: Nicolás Sandoval <nicolassandovalunm@gmail.com>

Remove Python 2.7 deprecated GitHub Action

Signed-off-by: Nicolás Sandoval <nicolassandovalunm@gmail.com>

refactor: removed repeated code in signing certificates

Signed-off-by: HanslettTheDev <HanslettTheDev@gmail.com>

HanslettTheDev committed:
Merge branch 'fix_signature' of https://github.com/HanslettTheDev/pyafipws into fix_signature

Signed-off-by: HanslettTheDev <HanslettTheDev@gmail.com>

HanslettTheDev committed:
refactoring the wssa.py file

Signed-off-by: HanslettTheDev <HanslettTheDev@gmail.com>
